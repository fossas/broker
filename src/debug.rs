//! Covers logging, tracing, metrics reporting, etc.
//!
//! # Artifacts
//!
//! Debug artifacts are generally similar to log files, but:
//! - They can contain more than just logs, for example traces.
//! - They can comprise other types of data, for example time series metrics snapshots.

use std::path::{Path, PathBuf};

use derive_more::{AsRef, From, Into};
use derive_new::new;
use error_stack::{report, Report};
use getset::{CopyGetters, Getters};
use rolling_file::{BasicRollingFileAppender, RollingConditionBasic};
use tracing::info;
use tracing_appender::non_blocking::WorkerGuard;
use tracing_subscriber::{filter, fmt::format::FmtSpan, prelude::*, Registry};

use crate::ext::{
    error_stack::{DescribeContext, ErrorHelper, IntoContext},
    result::WrapErr,
};

/// Errors that are possibly surfaced when running debugging operations.
#[derive(Debug, thiserror::Error)]
pub enum Error {
    /// When the trace sink is initialized, it is initialized as a global singleton.
    /// Future attempts to initialize it result in this error.
    /// This is a program logic error ("a bug"), and cannot be resolved by users.
    #[error("trace sink was configured again after being configured once")]
    TraceSinkReconfigured,

    /// When configuring tracing log output, it's possible for the rolling appender to fail.
    #[error("failed to configure tracing output location")]
    TraceConfig,

    /// When configuring tracing, we ensure that the tracing root directory exists.
    /// If it didn't exist and can't be created, this error is returned.
    #[error("failed to create tracing output location")]
    EnsureTraceRoot,
}

/// Errors that are possibly surfaced during validation of config values.
#[derive(Debug, thiserror::Error)]
pub enum ValidationError {
    /// Retentions must be above a minimum value.
    #[error("retention value is too small")]
    RetentionBelowMinimum,
}

/// Validated config values for observability.
#[derive(Debug, Clone, PartialEq, Eq, Getters, new)]
#[getset(get = "pub")]
pub struct Config {
    /// The location into which observability artifacts are stored.
    location: Root,

    /// The configured retention settings.
    retention: Retention,
}

impl Config {
    /// Initialize debugging singletons.
    ///
    /// Until this method is run, traces are not output anywhere and are lost forever;
    /// run it as soon as possible.
    #[must_use = "This guard must be stored in a variable that is retained; if it is dropped the tracing sink will stop running"]
    pub fn run_tracing_sink(&self) -> Result<WorkerGuard, Report<Error>> {
        self.ensure_tracing_root_exists()?;
        self.initialize_tracing_sink()
    }

    /// The path to the directory containing trace files.
    fn tracing_root(&self) -> PathBuf {
        self.location().as_ref().join("trace")
    }

    /// Ensure the tracing root exists.
    fn ensure_tracing_root_exists(&self) -> Result<(), Report<Error>> {
        let root = self.tracing_root();
        std::fs::create_dir_all(&root)
            .context(Error::EnsureTraceRoot)
            .help("this location is set in the config file")
            .describe_lazy(|| {
                format!(
                    "debug info is configured to be stored in {}",
                    root.display()
                )
            })
    }

    /// Initialize tracing sinks:
    /// - Hourly rotating sink of all raw traces in JSON format to disk.
    /// - Pretty sink of INFO-level traces to stdout.
    fn initialize_tracing_sink(&self) -> Result<WorkerGuard, Report<Error>> {
        let target = self.tracing_root().join("broker.trace");
        let file = self.retention().sink(&target)?;
        let (sink, guard) = tracing_appender::non_blocking(file);

        let subscriber = Registry::default()
            // log pretty info traces to terminal
            .with(
                tracing_subscriber::fmt::layer()
                    .pretty()
                    .with_filter(filter::LevelFilter::INFO),
            )
            // log all traces to file in json format
            .with(
                tracing_subscriber::fmt::layer()
                    .json()
                    .with_span_events(FmtSpan::FULL)
                    .with_writer(sink),
            );

        tracing::subscriber::set_global_default(subscriber)
            .context(Error::TraceSinkReconfigured)
            .help("if you're a user and you're seeing this, please report this as a defect to FOSSA support")
            .describe("this is a program bug and is not something that users can fix")?;

        info!(
            "Debug artifacts being stored in {}",
            self.tracing_root().display()
        );
        Ok(guard)
    }
}

/// Observability artifacts are stored on disk until requested by the FOSSA service.
/// This variable defines the root at which these artifacts are stored.
#[derive(Debug, Clone, PartialEq, Eq, AsRef, From, new)]
pub struct Root(PathBuf);

/// Since observability artifacts are stored on disk, we obviously want to clean them up.
/// These retention settings are used by a background process to keep artifact size in line.
#[derive(Debug, Clone, PartialEq, CopyGetters, Eq, new)]
#[getset(get_copy = "pub")]
pub struct Retention {
    /// The number of days to retain.
    days: ArtifactRetentionCount,
}

impl Retention {
    fn sink(&self, target: &Path) -> Result<BasicRollingFileAppender, Report<Error>> {
        let roll_condition = RollingConditionBasic::new().daily();
        BasicRollingFileAppender::new(target, roll_condition, self.days.into())
            .context(Error::TraceConfig)
            .help("ensure that the parent directory exists and you have access to it")
            .describe_lazy(|| format!("initialize sink to {}", target.display()))
    }
}

/// Specifies the number of rotated artifacts that are kept.
#[derive(Debug, Clone, Copy, PartialEq, Into, Eq, new)]
pub struct ArtifactRetentionCount(usize);

impl Default for ArtifactRetentionCount {
    /// Defaults to seven days.
    fn default() -> Self {
        Self(7)
    }
}

impl PartialEq<usize> for ArtifactRetentionCount {
    fn eq(&self, other: &usize) -> bool {
        self.0 == *other
    }
}

impl TryFrom<usize> for ArtifactRetentionCount {
    type Error = Report<ValidationError>;

    fn try_from(value: usize) -> Result<Self, Self::Error> {
        if value == 0 {
            report!(ValidationError::RetentionBelowMinimum)
                .wrap_err()
                .help("must specify at least '1'")
                .describe_lazy(|| format!("provided value: {value}"))
        } else {
            Ok(Self(value))
        }
    }
}
