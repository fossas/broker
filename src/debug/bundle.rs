//! The debug bundle is analogous to other FOSSA "debug bundle" implementations:
//!
//! - Debug bundle in FOSSA CLI
//! - Debug bundle in FOSSA Helm Charts
//!
//! The idea is that users can ask Broker to generate a "debug bundle",
//! which contains all the information FOSSA engineers need to reproduce or debug a problem,
//! while doing our best to not send internal information such as source code to FOSSA.

use std::path::{Path, PathBuf};

use error_stack::{Context, Result, ResultExt};
use getset::Getters;
use libflate::gzip;
use tempfile::NamedTempFile;
use thiserror::Error;
use tracing::{debug, error};
use walkdir::WalkDir;

use crate::{
    ext::{error_stack::IntoContext, io, tracing::span_records},
    AppContext,
};

/// Errors encountered generating a debug bundle.
#[derive(Debug, Error)]
pub enum Error {
    /// When Broker creates the debug bundle, it initially collects the debug artifacts in a temporary file.
    /// This is so that if the overall process fails Broker doesn't accidentally leave a partially-constructed
    /// debug bundle at the output location, potentially confusing the user.
    #[error("create temporary file")]
    CreateTempFile,

    /// When instantiating the bundler, if its options are not correct, this error is the result.
    #[error("create bundler")]
    CreateBundler,

    /// In order to write traces and FOSSA CLI debug bundles, Broker must be able to walk the debug directory.
    #[error("walk contents of directory: '{}'", .0.display())]
    WalkContents(PathBuf),

    /// Broker wasn't able to bundle the directory.
    #[error("bundle contents of file in directory '{}': '{}'", .dir.display(), .file.display())]
    BundleContents {
        /// The root directory for debug artifacts.
        dir: PathBuf,

        /// The file being bundled.
        file: PathBuf,
    },

    /// When finished bundling everything, it's finalized into a [`Bundle`].
    /// If this fails, this error is reported.
    #[error("finalize debug bundle")]
    Finalize,
}

impl Error {
    fn walk_contents(dir: &Path) -> Self {
        Self::WalkContents(dir.to_owned())
    }

    fn bundle_contents(dir: &Path, file: &Path) -> Self {
        Self::BundleContents {
            dir: dir.to_owned(),
            file: file.to_owned(),
        }
    }
}

/// Intended to provide FOSSA employees with all the information required to debug
/// Broker in any environment, without sharing sensitive information.
#[derive(Debug, Clone, Getters)]
#[getset(get = "pub")]
pub struct Bundle {
    /// The location to the debug bundle that was persisted.
    location: PathBuf,
}

/// Provider for writing files to a debug bundle.
pub trait Bundler {
    /// The error returned by this implementation.
    type Error;

    /// Add a new entry from a file on the local file system with the provided name.
    fn add_file<P, N>(&mut self, path: P, name: N) -> std::result::Result<(), Self::Error>
    where
        P: AsRef<Path>,
        N: AsRef<Path>;

    /// Finalize the bundle.
    fn finalize<P: AsRef<Path>>(self, destination: P) -> std::result::Result<Bundle, Self::Error>;
}

/// Metadata for a bundled entry.
pub struct EntryMetadata {}

/// Generate a debug bundle for the application at the specified location.
/// Ultimately this means "write a copy of every file inside the debug root to the bundler".
///
/// In the future, we'd like to decompress and potentially prettify the FOSSA CLI debug bundles
/// before including them in the overall debug bundle; this would yield better compression ratios.
/// For now though we just include them as is.
// #[tracing::instrument(skip(bundler, path), fields(debug_root, path))]
pub fn generate<B, P>(ctx: &AppContext, mut bundler: B, path: P) -> Result<Bundle, Error>
where
    B: Bundler,
    B::Error: Context,
    P: AsRef<Path>,
{
    let debug_root = ctx.data_dir("debug");
    span_records! {
        debug_root => display debug_root.display();
        path => display path.as_ref().display();
    }

    for entry in WalkDir::new(&debug_root).follow_links(false) {
        let entry = entry.context_lazy(|| Error::walk_contents(&debug_root))?;
        let path = entry.path();
        let rel = match path.strip_prefix(&debug_root) {
            Ok(rel) => rel,
            Err(err) => {
                error!(
                    debug_root = %debug_root.display(),
                    path = %path.display(),
                    %err,
                    "Skipping 'path': could not make relative to 'debug_root', see 'err' for details",
                );
                continue;
            }
        };

        if !entry.file_type().is_file() {
            debug!(path = %path.display(), "Skipping '{}': not a file", rel.display());
            continue;
        }

        // Copy the file to temp first, so that it's not changed while the tar is being built.
        let copy = io::sync::copy_temp(path).change_context(Error::CreateTempFile)?;
        bundler
            .add_file(copy.path(), rel)
            .context_lazy(|| Error::bundle_contents(&debug_root, rel))?;
    }

    bundler.finalize(path).context(Error::Finalize)
}

/// Implementation for [`Bundler`] which bundles files into a `.tar.gz` file.
pub struct TarGzBundler {
    inner: tar::Builder<gzip::Encoder<NamedTempFile>>,
}

impl TarGzBundler {
    /// Create a new instance.
    ///
    /// This bundles the provided debug artifacts using a backing temp file,
    /// which is moved to a final location with the `finalize` method.
    pub fn new() -> Result<Self, Error> {
        let file = NamedTempFile::new().context(Error::CreateTempFile)?;
        let encoder = gzip::Encoder::new(file).context(Error::CreateBundler)?;
        Ok(Self {
            inner: tar::Builder::new(encoder),
        })
    }
}

impl Bundler for TarGzBundler {
    type Error = std::io::Error;

    fn add_file<P, N>(&mut self, path: P, name: N) -> std::result::Result<(), Self::Error>
    where
        P: AsRef<Path>,
        N: AsRef<Path>,
    {
        self.inner.append_path_with_name(path, name)
    }

    fn finalize<P: AsRef<Path>>(self, destination: P) -> std::result::Result<Bundle, Self::Error> {
        let zip = self.inner.into_inner()?;
        let handle = zip.finish().into_result()?;
        handle.as_file().sync_all()?;
        handle.persist(&destination)?;
        Ok(Bundle {
            location: destination.as_ref().to_path_buf(),
        })
    }
}
