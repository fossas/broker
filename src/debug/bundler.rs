//! Trait and implementations used by `bundle` to create debug bundles.

use std::{fs, path::Path};

use error_stack::Result;
use libflate::gzip;
use tempfile::NamedTempFile;
use thiserror::Error;

use crate::ext::error_stack::IntoContext;

use super::Bundle;

/// Errors encountered generating a debug bundle.
#[derive(Debug, Error)]
pub enum Error {
    /// When Broker creates the debug bundle, it initially collects the debug artifacts in a temporary file.
    /// This is so that if the overall process fails Broker doesn't accidentally leave a partially-constructed
    /// debug bundle at the output location, potentially confusing the user.
    #[error("create temporary file")]
    CreateTempFile,

    /// When instantiating the bundler, if its options are not correct, this error is the result.
    #[error("create bundler")]
    CreateBundler,
}

/// Provider for writing files to a debug bundle.
pub trait Bundler {
    /// The error returned by this implementation.
    type Error;

    /// Add a new entry from a file on the local file system with the provided name.
    fn add_file<P, N>(&mut self, path: P, name: N) -> std::result::Result<(), Self::Error>
    where
        P: AsRef<Path>,
        N: AsRef<Path>;

    /// Finalize the bundle.
    fn finalize<P: AsRef<Path>>(self, destination: P) -> std::result::Result<Bundle, Self::Error>;
}

/// Implementation for [`Bundler`] which bundles files into a `.tar.gz` file.
pub struct TarGz {
    inner: tar::Builder<gzip::Encoder<NamedTempFile>>,
}

impl TarGz {
    /// Create a new instance.
    ///
    /// This bundles the provided debug artifacts using a backing temp file,
    /// which is moved to a final location with the `finalize` method.
    #[tracing::instrument]
    pub fn new() -> Result<Self, Error> {
        let file = NamedTempFile::new().context(Error::CreateTempFile)?;
        let encoder = gzip::Encoder::new(file).context(Error::CreateBundler)?;
        Ok(Self {
            inner: tar::Builder::new(encoder),
        })
    }
}

impl Bundler for TarGz {
    type Error = std::io::Error;

    #[tracing::instrument(skip_all, fields(path = %path.as_ref().display(), name = %name.as_ref().display()))]
    fn add_file<P, N>(&mut self, path: P, name: N) -> std::result::Result<(), Self::Error>
    where
        P: AsRef<Path>,
        N: AsRef<Path>,
    {
        self.inner.append_path_with_name(path, name)
    }

    #[tracing::instrument(skip_all, fields(destination = %destination.as_ref().display()))]
    fn finalize<P: AsRef<Path>>(self, destination: P) -> std::result::Result<Bundle, Self::Error> {
        let zip = self.inner.into_inner()?;
        let handle = zip.finish().into_result()?;
        handle.as_file().sync_all()?;

        // `handle.persist` fails if asked to persist across mounts, since internally it uses a rename.
        // Instead, just copy the file from temp.
        let path = handle.into_temp_path();
        fs::copy(&path, destination.as_ref())?;

        Ok(Bundle {
            location: destination.as_ref().to_path_buf(),
        })
    }
}
