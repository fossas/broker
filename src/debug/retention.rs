//! Responsible for cleaning up tracing files (and maybe other artifacts in the future).
//!
//! # Tracing files
//!
//! Stored in `self.location().join("trace")`. Each hour, a file is rotated;
//! these are stored in the format `broker.trace.yyyy-MM-dd-HH`,
//! where the timestamp is the UTC time.
//!
//! So we just need to find files with timestamps older than the configured timeout.

use std::{
    cmp::Ordering,
    fmt::{Debug, Display},
    fs::Metadata,
    ops::Add,
    os::unix::prelude::MetadataExt,
    path::Path,
    time::{Duration, SystemTime},
};

use async_stream::try_stream;
use bytesize::ByteSize;
use delegate::delegate;
use derive_more::From;
use derive_new::new;
use error_stack::{report, Report, ResultExt};
use futures::{Stream, StreamExt, TryStreamExt};
use tokio::{
    fs::{self, DirEntry},
    time::{self, Instant},
};
use tracing::{debug, info, warn};

use crate::ext::{
    error_stack::{merge_error_stacks, DescribeContext, ErrorHelper, IntoContext},
    result::InspectErr,
};

use super::Config;

const WORKER_CYCLE_DELAY: Duration = Duration::from_secs(60);

/// Errors that are possibly surfaced when running debugging operations.
#[derive(Debug, thiserror::Error)]
pub enum Error {
    /// Encountered cleaning up max age.
    #[error("retention by age")]
    ByAge,

    /// Encountered cleaning up max size.
    #[error("retention by size")]
    BySize,

    /// Encounted while enumerating files.
    #[error("enumerate files")]
    EnumerateFiles,

    /// Encountered while removing a file.
    #[error("remove file")]
    RemoveFile,
}

/// Run the tracing cleanup worker.
///
/// This runs as an async background task, meaning that it needs to be constantly driven.
/// If it is dropped at any point, the worker stops.
pub async fn run_worker(config: &Config) -> Result<(), Report<Error>> {
    loop {
        let artifacts = do_cleanup_cycle(config)
            .await
            .describe("perform cleanup cycle")?;

        if !artifacts.is_empty() {
            info!("Cleaned up artifacts:\n {artifacts}");
        }

        let next_cycle = Instant::now().add(WORKER_CYCLE_DELAY);
        debug!("Running next cleanup at {next_cycle:?}");
        time::sleep_until(next_cycle).await;
    }
}

/// The list of cleaned up artifacts in a cleanup cycle.
#[derive(Debug, Clone, PartialEq, Eq, From, Default, new)]
struct CleanedUpArtifacts(Vec<String>);

impl CleanedUpArtifacts {
    delegate! {
        to self.0 {
            /// The number of artifacts that were cleaned up.
            pub fn len(&self) -> usize;

            /// Check whether any artifacts were cleaned up.
            pub fn is_empty(&self) -> bool;

            /// Iterate over the artifacts that were cleaned up.
            pub fn iter(&self) -> impl Iterator<Item = &String>;
        }
    }
}

impl Add for CleanedUpArtifacts {
    type Output = CleanedUpArtifacts;

    fn add(mut self, rhs: Self) -> Self::Output {
        self.0.extend_from_slice(&rhs.0);
        self
    }
}

impl Display for CleanedUpArtifacts {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let pretty = self
            .iter()
            .map(|artifact| format!("\t{artifact}"))
            .collect::<Vec<_>>()
            .join("\n");
        write!(f, "{pretty}")
    }
}

/// Run a cleanup cycle.
#[tracing::instrument]
async fn do_cleanup_cycle(config: &Config) -> Result<CleanedUpArtifacts, Report<Error>> {
    // These two cleanups should run serially so that they don't step on each other's toes.
    let age = do_cleanup_cycle_age(config)
        .await
        .change_context(Error::ByAge);
    let size = do_cleanup_cycle_size(config)
        .await
        .change_context(Error::BySize);

    match (age, size) {
        (Ok(age), Ok(size)) => Ok(age + size),
        (Err(age), Err(size)) => Err(merge_error_stacks!(age, size)),
        (Ok(age), Err(size)) => Err(report!(size)).help(format!(
            "age retention ran successfully, cleaning up {} items",
            age.len()
        )),
        (Err(age), Ok(size)) => Err(report!(age)).help(format!(
            "size retention ran successfully, cleaning up {} items",
            size.len()
        )),
    }
}

#[tracing::instrument]
async fn do_cleanup_cycle_age(config: &Config) -> Result<CleanedUpArtifacts, Report<Error>> {
    let Some(max_age) = config.retention().age() else { return Ok(CleanedUpArtifacts::default()) };

    list_contents(&config.tracing_root())
        .try_filter_map(|entry| async move {
            let meta = entry
                .metadata()
                .await
                .context(Error::EnumerateFiles)
                .describe_lazy(|| format!("list metadata for file {:?}", entry.path()))?;
            if meta.is_file() && max_age.is_violated_by(file_age(&meta)) {
                Ok(Some(entry.path()))
            } else {
                Ok(None)
            }
        })
        .and_then(|path| async move {
            fs::remove_file(&path)
                .await
                .context(Error::RemoveFile)
                .describe_lazy(|| format!("remove file {path:?}"))
                .map(|_| path.to_string_lossy().to_string())
        })
        .try_collect::<Vec<_>>()
        .await
        .map(CleanedUpArtifacts::from)
        .help("ensure the file system is not read only")
}

/// Clean up the oldest files that are causing the total file size to exceed the max retention size.
#[tracing::instrument]
async fn do_cleanup_cycle_size(config: &Config) -> Result<CleanedUpArtifacts, Report<Error>> {
    let Some(max_size) = config.retention().size() else { return Ok(CleanedUpArtifacts::default()) };

    // It's not efficient to sort streams, so collect into a vector to do the sorting.
    let mut files = list_contents(&config.tracing_root())
        .try_filter_map(|entry| async move {
            let meta = entry
                .metadata()
                .await
                .context(Error::EnumerateFiles)
                .describe_lazy(|| format!("list metadata for file {:?}", entry.path()))?;

            if meta.is_file() {
                let path = entry.path();
                let age = file_age(&meta);
                let size = ByteSize::b(meta.size());
                Ok(Some((age, size, path)))
            } else {
                Ok(None)
            }
        })
        .try_collect::<Vec<_>>()
        .await?;

    // If two files are the exact same age, use path to break the tie.
    files.sort_unstable_by(
        |(age_a, _, path_a), (age_b, _, path_b)| match age_a.cmp(age_b) {
            Ordering::Equal => path_a.cmp(path_b),
            other => other,
        },
    );

    // Now walk through the sorted vec, deleting the ones that cause the overall collection to go over the max size.
    let mut total_size = ByteSize::b(0);
    futures::stream::iter(files)
        .filter_map(|(_, size, path)| async move {
            total_size += size;
            if max_size.is_violated_by(total_size) {
                debug!("{path:?} is over overall size limit: {max_size}");
                Some(path)
            } else {
                None
            }
        })
        .then(|path| async move {
            fs::remove_file(&path)
                .await
                .context(Error::RemoveFile)
                .describe_lazy(|| format!("remove file {path:?}"))
                .map(|_| path.to_string_lossy().to_string())
        })
        .try_collect::<Vec<_>>()
        .await
        .map(CleanedUpArtifacts::from)
        .help("ensure the file system is not read only")
}

/// Create a stream which lists the contents of a single level of a directory.
fn list_contents(dir: &Path) -> impl Stream<Item = Result<DirEntry, Report<Error>>> {
    // Take ownership of the path (by copying it) so that we can send it across threads without having to manage external lifetimes.
    let dir = dir.to_path_buf();

    // Implements a stream (basically an async iterator) via `await` and `yield`.
    // Once this block exits, the stream is closed.
    try_stream! {
        let mut contents = fs::read_dir(&dir)
            .await
            .context(Error::EnumerateFiles)
            .help("ensure the current user has access to list files in the directory")
            .describe_lazy(|| format!("list contents of {dir:?}"))?;

        loop {
            // These error helpers are the same basically just because there's no useful additional information to give in this case.
            let file = contents.next_entry().await
                .context(Error::EnumerateFiles)
                .help("ensure the current user has access to list files in the directory")
                .describe_lazy(|| format!("list contents of {dir:?}"))?;

            // Yield files in the loop until the underlying iterator is done.
            let Some(file) = file else { break };
            yield file;
        }
    }
}

/// Returns the canonical age of a file.
///
/// Uses the file's last modified time (if possible, falling back to the creation time)
/// and then compares that to the current system time.
///
/// If anything in that chain fails, one or more warnings are emitted
/// and the file age is assumed to be "zero" (meaning it is assumed to have just been created).
///
/// # Failures
///
/// File modification time can fail if the system does not provide a file modification time API,
/// or if the file system does not support modification times.
///
/// File creation time can fail if the system does not provide a file modification time API,
/// or if the file system does not support creation times.
///
/// File age comparison fails if _both_ of the above fail, _or_ if the current system time
/// is earlier than the canonical modification time
/// (this usually is due to the system clock being messed with).
///
/// In all failure cases, the file's age is assumed to be zero.
fn file_age(metadata: &Metadata) -> Duration {
    metadata
        .modified()
        .ext_inspect_err(|err| warn!("file modified time not available on platform: {err}"))
        .or_else(|_| metadata.created())
        .ext_inspect_err(|err| warn!("file created time not available on platform: {err}"))
        .ok()
        .and_then(|modified| {
            let now = SystemTime::now();
            let elapsed = now.duration_since(modified);
            elapsed.ext_inspect_err(|err| warn!("unable to determine elapsed time between {modified:?} and {now:?} (possibly due to system clock drift): {err}")).ok()
        })
        .unwrap_or(Duration::ZERO)
}
