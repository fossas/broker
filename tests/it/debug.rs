//! Tests for debugging functionality.

use broker::debug::ArtifactRetentionCount;
use proptest::prelude::*;
use test_strategy::proptest;

use crate::helper::{assert_error_stack_snapshot, load_config};

#[proptest]
fn validate_artifact_retention_count(
    #[by_ref]
    #[filter(*#input > 0)]
    input: usize,
) {
    match ArtifactRetentionCount::try_from(input) {
        Ok(validated) => prop_assert_eq!(validated, input, "tested input: {:?}", input),
        Err(err) => prop_assert!(false, "unexpected parsing error '{err:#}' for '{input}'"),
    }
}

#[test]
fn validate_artifact_retention_count_min() {
    let input = 0;
    assert_error_stack_snapshot!(
        &input,
        ArtifactRetentionCount::try_from(input).expect_err("must have failed validation")
    )
}

#[test]
fn validate_artifact_retention_count_default() {
    assert_eq!(ArtifactRetentionCount::default(), 7);
}

#[tokio::test]
async fn test_debug_location_invalid() {
    let (config_path, config) = load_config!(
        "testdata/config/basic-location-invalid.yml",
        "testdata/database/empty.sqlite"
    )
    .await;

    let err = config
        .debug()
        .run_tracing_sink()
        .expect_err("must have errored");
    assert_error_stack_snapshot!(&config_path, err);
}
