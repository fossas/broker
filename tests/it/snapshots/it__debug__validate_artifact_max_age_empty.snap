---
source: tests/it/debug.rs
expression: "ArtifactMaxAge::try_from(input).expect_err(\"must have failed validation\")"
info: ""
---
validate retention age
├╴at {source location}
├╴context: provided value: 
│
├─▶ parse duration
│   ╰╴at {source location}
│
╰─▶ value was empty
    ╰╴at {source location}
